src/dinit.cc

```diff
@@ -696,6 +700,15 @@ int dinit_main(int argc, char **argv)
     close_control_socket();
     
     if (am_system_mgr) {
+        if (shutdown_type == shutdown_type_t::SOFTREBOOT) {
+            sync(); // Sync to minimise data loss if user elects to power off / hard reset
+            execv(dinit_exec, argv);
```

Comment:
(I just noticed this PR is still marked as WIP. Apologies if you hadn't finished yet - I will wait until you remove the WIP status before I review any further, but in general it looks ok otherwise).

Reply:
I thought of this and have several possible ideas.

1. strip the `single` argument from the `argv`, which works, but the strip process is complicated and requires more modification if new acceptable kernel arguments are added subsequently.
2. set a new cmdline argument `--no-process-sys-args`, and place it at the beginning of the `argv` when re-`execv`-ing.
3. add a new argument to `shutdown` that overrides the `argv` when re-`execv`-ing.

I think this needs further discussion.

Reply:
Added an cmdline arg `--ignore-sys-args` to ignore kernel args like `single`